### PR TITLE
packet peek

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
add ability to peek ahead in packets but remain immutable as we parse

- adds an immutable smart pointer that prevents a packet from being modified. 
- adds a *conditional* (via enum) reference counted smart pointer allowing us to use raw (for parse) or reference counted (for peek) envelopes.

### CODE REVIEW

File: framework/src/control/tcp.rs [hunk 1]
```diff
@@ -96,7 +96,6 @@ impl<T: TcpControlAgent> TcpControlServer<T> {
                 );
             // Add to some sort of hashmap.
             } else {
-
```
Comment (zeeshanlakhani) [inline framework/src/control/tcp.rs]: just running `make fmt`

File: framework/src/packets/mod.rs [hunk 2]
```diff
@@ -154,6 +173,92 @@ pub trait Packet {
...
+    Raw(T),
+    Counted(Rc<T>),
+}
+
+impl<T: Packet> CondRc<T> {
+    pub fn new(value: T) -> Self {
+        CondRc::Raw(value)
+    }
+
+    pub fn into_owned(self) -> T {
```
Comment (zeeshanlakhani) [inline framework/src/packets/mod.rs]: @drunkirishcoder into_owned b/c `to_*` functions usually take in references (via clippy docs), where this is taking in our non-referenced enum and unwrapping it.

File: framework/src/packets/ethernet.rs [hunk 3]
```diff
@@ -385,4 +397,16 @@ mod tests {
...
+    #[dpdk_test]
+    fn peek_v4_udp_packet() {
+        let packet = RawPacket::from_bytes(&UDP_PACKET).unwrap();
+        let ethernet = packet.parse::<Ethernet>().unwrap();
+
+        let v4 = ethernet.peek::<Ipv4>().unwrap();
+        assert_eq!(255, v4.ttl());
+        assert_eq!(MacAddr::new(0, 0, 0, 0, 0, 2), v4.envelope().src());
+
+        let udp = v4.peek::<Udp<Ipv4>>().unwrap();
```
Comment (drunkirishcoder) [inline framework/src/packets/ethernet.rs]: 👏 very cool, it works.

File: framework/src/packets/mod.rs (hunk 2 above)
Comment (drunkirishcoder) [reply framework/src/packets/mod.rs]: yep `into_*` is the right convention here, good call.

**Comment by drunkirishcoder** [comment]
Nevermind, I forgot `filter` and `group_by` don't allow parsing already. Now with this you can `peek` into the payload. no additional PR needed.

**Pull request merged**